cmd/crank/pkg/build.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return err
+	}
+
+	_, err = io.Copy(b.logWriter, resp)
+	if err != nil {
+		return err
+	}
+	return resp.Close()
+}
+
+func (b *buildImage) templateDockerfile() (string, error) {
+	tmpl := `
+FROM {{.BaseImage}}
+
+# We assume that WORKDIR has been set by the base image
+# to control where files will be placed by docker when we specify '.'/CWD
+ADD {{.PackageContentsPath}} .
+`
+	t := template.Must(template.New("dockerfile").Parse(tmpl))
+	buf := new(bytes.Buffer)
+	values := struct {
+		BaseImage           string
+		PackageContentsPath string
+	}{
+		BaseImage:           b.baseImage,
+		PackageContentsPath: DockerTarRoot,
+	}
+	err := t.Execute(buf, values)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+// dockerBuild invokes docker's /build api
+// https://docs.docker.com/engine/api/v1.40/#operation/ImageBuild
+func (b *buildImage) dockerBuild(ctx context.Context, dockerfile string) error {
+	cops, err := b.copyOperations()
+	if err != nil {
+		return err
+	}
+
+	tb, err := b.buildTar(dockerfile, cops)
+	if err != nil {
+		return err
+	}
+
+	// not setting NoCache or PullParent to make local development workflows easier
+	// TODO: do we need to expose these as flags or just leave them off (default)
+	opts := types.ImageBuildOptions{
+		Tags: []string{b.pkgImageName},
+	}
+	resp, err := b.docker.ImageBuild(ctx, tb, opts)
+	if err != nil {
+		return err
+	}
+
+	buf := new(bytes.Buffer)
+	_, err = io.Copy(buf, resp.Body)
+	_ = resp.Body.Close()
+	if err != nil {
+		return err
+	}
+	fmt.Println(buf.String())
+
+	return nil
+}
+
+func (b *buildImage) buildTar(dockerfile string, cops []copyOperation) (io.Reader, error) {
+	buf := new(bytes.Buffer)
+	tw := tar.NewWriter(buf)
+
+	// Start by writing the Dockerfile in at the root
+	dhdr := &tar.Header{
+		Name: "Dockerfile",
+		Mode: 0600,
+		Size: int64(len(dockerfile)),
+	}
+	if err := tw.WriteHeader(dhdr); err != nil {
+		return nil, err
+	}
+	_, err := io.Copy(tw, bytes.NewBufferString(dockerfile))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, co := range cops {
+		fi, err := os.Stat(co.srcPath)
+		if err != nil {
+			return nil, err
+		}
+		fh, err := os.Open(co.srcPath)
+		if err != nil {
+			return nil, err
+		}
+		hdr := &tar.Header{
+			Name: co.destPath,
+			Mode: int64(fi.Mode()),
+			Size: fi.Size(),
+		}
+		if err := tw.WriteHeader(hdr); err != nil {
+			return nil, err
+		}
+		_, err = io.Copy(tw, fh)
+		if err != nil {
+			return nil, err
+		}
+		// lint made me do it
+		_ = fh.Close()
+	}
+	err = tw.Close()
+	return buf, err
+}
+
+func (b *buildImage) copyOperations() ([]copyOperation, error) {
+	cops := make([]copyOperation, 1)
+	cops[0] = copyOperation{
+		srcPath:  filepath.Join(b.basePath, "crossplane.yaml"),
+		destPath: filepath.Join(DockerTarRoot, "crossplane.yaml"),
+	}
+	err := filepath.Walk(b.basePath, func(path string, info os.FileInfo, err error) error {
```

Comment:
Wonder if this could be an opportunity to use some of the logic implemented here: https://github.com/crossplane/crossplane-runtime/blob/e1fc0493b2e88dc4b067634eed445cda345a9843/pkg/parser/fsreader.go#L72

Reply:
What I really need is the set of filtered paths and a `Reader` for each so that I can build up the docker tarball. The paths are private in the `FsReadCloser` type and the reader returns a stream with all contents concatenated, so I think it isn't a good fit.